### ISSUE TITLE
properly handle 4xx status codes

### ISSUE THREAD

**Issue by ivantopo** [kamon-io/kamon-akka-http#29]
We have this little piece of code in the [server side instrumentation](https://github.com/kamon-io/kamon-akka-http/blob/master/kamon-akka-http/src/main/scala/kamon/akka/http/instrumentation/ServerFlowWrapper.scala#L88):

```
          if(status >= 400 && status <= 499) {
            span.setOperationName("not-found")
          }
```

That is definitely wrong, but I'm not sure of what to do about it :confused:.. the motivation behind this was to prevent requests to unknown resources from generating lots of metrics in Kamon since  the operation name is used as a metric tag and anything left out could lead to cardinality explosion. This solution is too conservative, though.

@Falmarri brought this up on [Gitter](https://gitter.im/kamon-io/Kamon?at=5a6413075ade18be39a179ab), but not the first time this shows up in the last few days.. @ptrlaszlo recently [proposed a PR](https://github.com/kamon-io/kamon-akka-http/pull/28) that I think is somehow related to this and @n1ko-w1ll also gave me some feedback regarding this being sort of a problem so, your inputs on how we could make this handling more robust and still keep cardinality at bay  would be greatly appreciated!

In a ideal integration all routes recognized by the application will already have a stable name provided by either a name generator or the `operationName` directive.. maybe the best course of action would be to have a rejection handler that will turn everything else into a `unhandled` operation name and additionally be able to automatically add the status code as a metric tag via config option as @ptrlaszlo proposed.

How does that sound to you? /cc @dpsoft @mladens @jypma, would like to hear your comments as well! :)

**Comment by Falmarri** [comment]
If you want to handle the not found specifically, you can specifically check for 404, i would be reasonably ok with that. Or you can inject higher into the DSL to look for rejected route with an empty rejection list (which indicates 404). Or keep track if the operation name is overridden or not and only set it if it is.

Another option is to set the operation name at the beginning to a sentinel (eg "root", which is what i use), and then before finishing, rename it with the path if the operation is still that sentinel value.

But regardless, you certainly need to let other 4xx be recorded as errors with their appropriate operation name

>since the operation name is used as a metric tag and anything left out could lead to cardinality explosion

I think this is a flaw of the default generator imo.

>have a rejection handler that will turn everything else into a unhandled operation

Perfectly fine with me. I'd like this to be either configurable or at least include the path as a tag. That's currently what I do

**Comment by n1ko-w1ll** [comment]
> In a ideal integration all routes recognized by the application will already have a stable name provided by either a name generator or the operationName directive.. maybe the best course of action would be to have a rejection handler that will turn everything else into a unhandled operation name and additionally be able to automatically add the status code as a metric tag via config option as @ptrlaszlo proposed.

I'm no expert with regards to Kamon or akka-http routes, this however sounds reasonable to me. In the end every solution is perfectly fine for me as long as I can measure how often a specific operation resulted in a 200 / 400 / 500 or whatever status code it might be.

**Comment by ivantopo** [comment]
I think I mixed a couple concerns on this issue so I'm splitting it in two:

  - Fixing the particular bug of putting all 4xx status codes on the same bucket, addressed by https://github.com/kamon-io/kamon-akka-http/pull/32 and will be released as 1.0.1 asap. Comment on the PR if there is anything weird.
  - Finding a solid solution to the cardinality issues, we will track that as https://github.com/kamon-io/kamon-akka-http/issues/31. This is a much bigger goal and will definitely not be a quick one, please dump any additional feedback related to this into that issue.
